astropy/io/ascii/cparser.pyx

```diff
@@ -343,7 +341,12 @@ cdef class CParser:
                 raise core.InconsistentTableError('No data lines found, C reader '
                                             'cannot autogenerate column names')
             # auto-generate names
-            self.names = ['col{0}'.format(i + 1) for i in range(self.width)]
+            self.header_names = ['col{0}'.format(i + 1) for i in range(self.width)]
+
+        if self.names:
+            self.width = len(self.names)
+        else:
+            self.names = self.header_names
```

Comment:
Yeah, the logic could definitely be tightened up here.  This was a result of trying to make a minimal change to the existing flow.  Given the crunch to try getting PRs merged today I'll leave it as-is but plan to get back to it later.